File: internal/config/format.go
package config

import "io"

type Format interface {
	Name() string
	Extensions() []string
	Parse(r io.Reader) (*Kylefile, error)
}

var registry = make(map[string]Format)

func Register(f Format) {
	registry[f.Name()] = f
}

func GetFormat(name string) (Format, bool) {
	f, ok := registry[name]
	return f, ok
}

func GetFormatByExtension(ext string) (Format, bool) {
	for _, f := range registry {
		for _, e := range f.Extensions() {
			if e == ext {
				return f, true
			}
		}
	}
	return nil, false
}

func ListFormats() []string {
	names := make([]string, 0, len(registry))
	for name := range registry {
		names = append(names, name)
	}
	return names
}

File: src/config/format.rs
use super::Error;
use super::kylefile::Kylefile;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Format {
    Yaml,
    Toml,
}

impl Format {
    pub fn name(self) -> &'static str {
        match self {
            Self::Yaml => "yaml",
            Self::Toml => "toml",
        }
    }

    pub fn extensions(self) -> &'static [&'static str] {
        match self {
            Self::Yaml => &[".yaml", ".yml"],
            Self::Toml => &[".toml"],
        }
    }

    pub fn from_name(name: &str) -> Option<Self> {
        if name.eq_ignore_ascii_case("yaml") {
            Some(Self::Yaml)
        } else if name.eq_ignore_ascii_case("toml") {
            Some(Self::Toml)
        } else {
            None
        }
    }

    pub fn from_extension(ext: &str) -> Option<Self> {
        match ext {
            ".yaml" | ".yml" => Some(Self::Yaml),
            ".toml" => Some(Self::Toml),
            _ => None,
        }
    }

    pub fn parse(self, content: &str) -> Result<Kylefile, Error> {
        match self {
            Self::Yaml => serde_yml::from_str(content).map_err(Error::Yaml),
            Self::Toml => toml::from_str(content).map_err(Error::Toml),
        }
    }
}
